_posts/2021-05-18-row_pattern_matching.md

```diff
@@ -0,0 +1,474 @@
+            PATTERN (START DOWN+ UP+)
+            DEFINE
+                DOWN AS price < PREV(price),
+                UP AS price > PREV(price)
+            );
+
+ customer_id | start_price | bottom_price | final_price | start_date | final_date
+-------------+-------------+--------------+-------------+------------+------------
+ cust_1      |         200 |           50 |         100 | 2020-05-12 | 2020-05-17
+ cust_2      |           8 |            4 |           6 | 2020-05-13 | 2020-05-18
+(2 rows)
+``` 
+
+Two matches are detected, one for `cust_1`, and one for `cust_2`.
+
+## Empty matches explained
+
+An empty match is a legit result of row pattern recognition. There are
+different pattern constructs that can result in an empty match. The empty
+pattern syntax `()` is the trivial one. Empty match can also result e.g. from
+quantification: `A*`, or alternation: `A | ()`.
+
+An empty match does not consume any input rows, but like every match, it is
+associated with a row, called the _starting row_. That is the row at which the
+pattern matching started. Note that if the pattern allows an empty match, it
+guarantees that no rows remain unmatched. Also, an empty match, as well as
+non-empty matches, gets a sequential number, which can be retrieved by the
+`MATCH_NUMBER` function.
+
+Depending on your use case, you can consider empty matches informative or just
+see them as a leftover of the algorithm.
+
+There's one more thing linked to empty matches. Some patterns have the
+dangerous potential of looping endlessly over a piece that doesn't consume any
+rows. It doesn't have to be as explicit as `()*`. There are complex patterns
+that don't show their looping potential at first glance. We handled them
+carefully so that you never have to waste your time on looping queries. 
+
+## In a few words, what's so cool about row pattern matching?
+
+From the SQL viewpoint, you can think of row pattern matching as extended
+window functions. Window functions allow you to capture some dependencies in
+rows of data based on their relative position or value. Row pattern matching
+allows you to detect arbitrarily complicated dependencies, based not only on
+the input values but also on the details of the actual match and on the match
+number.
+
+Before the introduction of `MATCH_RECOGNIZE`, you had to feed your data to
+external tools to reason about trends and patterns. Now, you can achieve it
+directly in your query, and even build your query upon the pattern recognition
+clause to further process the match results.
+
+Row pattern matching is typically used:
+
+- in trade applications for tracking trends or identifying customers with
+specific behavioral patterns,
+ 
+- in shipping applications for tracking packages through all possible valid
+paths,
+
+- in financial applications for detecting unusual incidents, which might signal
+fraud.
+
+What's your use case?
+
+I hope you enjoy Trino's new feature. [Please **do** reach out to us with any
```

Comment:
I hope you enjoy Trino's new feature. Refer to your documentation for even more details, and reach to us with. ..